src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -480,19 +480,18 @@ APPLICATION_MANAGER::DoRecycleApplication(
 // static
 VOID
 APPLICATION_MANAGER::RecycleApplication(
-    _In_ APPLICATION_INFO *     pEntry,
-    _In_ APP_HOSTING_MODEL      hostingModel
+    _In_ APPLICATION_INFO *     pEntry
 )
 {
-
     APPLICATION* pApplication = pEntry->QueryApplication();
+    DBG_ASSERT(pApplication != NULL);
 
     // Reference the application first
     pApplication->ReferenceApplication();
 
-    if (hostingModel == APP_HOSTING_MODEL::HOSTING_OUT_PROCESS)
+    if (pApplication->QueryConfig()->QueryHostingModel() == HOSTING_OUT_PROCESS)
     {
-        pEntry->ClearApplication();
+        pEntry->ClearAndDereferenceApplication();
```

Comment:
Are we confident that we can dereference the application too here?

Reply:
yes, as we reference it first, the reference counter become 2, we need dereference it to reduce the counter to 1 and then deference after recycle. The application then gets deleted.

Reply:
sounds good.